mirgecom/mpi.py

```diff
@@ -287,12 +283,22 @@ def wrapped_func(*args, **kwargs) -> None:
         # exit
         from mpi4py import MPI  # noqa
 
-        _check_gpu_oversubscription()
-        _check_cache_dirs_node()
+        from mirgecom.array_context import initialize_actx
+        # This is where actual actx selection occurs
+        actx = initialize_actx(actx_class, comm=MPI.COMM_WORLD)
+        kwargs["actx"] = actx
```

Comment:
This kind of worries me, but I'm not sure there is a better way to do this.

Reply:
I'm getting the feeling that `mpi_entry_point` has outlived its usefulness... I wonder if it would be cleaner at this point to just split this into two function calls, e.g., add an `initialize_mpi` function and then put the rest inside `initialize_actx`?

Reply:
Good idea, the only snag is that `initialize_mpi` also needs to know the actx_class, but not a big deal. What do you think of [bf9acf4](https://github.com/illinois-ceesd/mirgecom/pull/1030/commits/bf9acf4a02118cd938581f273c8694cbc3793494).

Reply:
Hmm, I forgot about that. Passing `actx_class` to `initialize_mpi` seems a little odd. Would it make sense to rename it to something like `initialize_env` instead and call it unconditionally from the driver, e.g.:

```python
from mirgecom.array_context import actx_class_is_pyopencl
initialize_env(use_cl=actx_class_is_pyopencl(actx_class), use_mpi=mpi)
```

?

Reply:
> ```python
> from mirgecom.array_context import actx_class_is_pyopencl
> initialize_env(use_cl=actx_class_is_pyopencl(actx_class), use_mpi=mpi)
> ```
I think this is also a bit clumsy, what do you think of [e833415](https://github.com/illinois-ceesd/mirgecom/pull/1030/commits/e8334150f37ea1b50bce12ecaceae99f84df9da9)?

Reply:
Ehh, I dunno if I like that... it imposes hidden constraints on how the driver creator can set up their command line arguments. Better to be explicit IMO.

Reply:
I dunno. The only other option that I can think of is to go back to the previous code in this location, i.e. running `cl.get_platforms()` unconditionally.

Note that the particular code here is almost completely irrelevant - the only reason for not running `cl.get_platforms()` in a numpy actx case is to avoid *pocl* from getting loaded, pyopencl will be loaded in for all actxs. There is no downside in running `cl.get_platforms()` (e.g., it will not ask for device selection or create a context). There is also no downside in loading pocl unncessarily.

Reply:
How about
```diff
-    if args.mpi:
-        main_func = main
-    else:
-        import inspect
-        # run main without the mpi_entry_point wrapper
-        main_func = inspect.unwrap(main)    if args.mpi:
+    if args.mpi:
+        main_func = mpi_entry_point(using_cl=actx_class_is_pyopencl(actx_class))(main)
+    else:
+        main_func = main
```
?

Reply:
I find it clumsy, and every driver needs to be updated to support it ...

Reply:
[5aac0c8](https://github.com/illinois-ceesd/mirgecom/pull/1030/commits/5aac0c81bda420c4dc67b001747538ca201112ed) (unconditional loading) is OK with me if there's no real downside. @MTCam?